workspaces/adventure-pack/goodies/typescript/Iterator.prototype.flatMap/index.ts

```diff
@@ -0,0 +1,34 @@
+import "../Iterator.prototype.toIterable";
+import { iteratorPrototype } from "../Iterator.prototype";
+
+declare global {
+  interface Iterator<T> {
+    flatMap<TOut>(
+      this: Iterator<T>,
+      callbackFn: (element: T, index: number) => Iterator<TOut>,
+    ): Generator<TOut, void, void>;
+  }
+}
+
+iteratorPrototype.flatMap ??= function* <TIn, TOut>(
+  this: Iterator<TIn>,
+  callbackFn: (element: TIn, index: number) => Iterator<TOut>,
+): Generator<TOut, void, void> {
+  if (typeof this !== "object") {
+    throw new TypeError("flatMap called on non-object");
+  }
+
+  let index = 0;
+  for (const element of this.toIterable()) {
+    const mapped = callbackFn(element, index).toIterable();
+    if (typeof mapped[Symbol.iterator] !== "function") {
```

Comment:
So I think that if the callback doesn't return an iterator or an iterable, then we'll already have an error at line 23 when trying to invoke the `.toIterable()` method. Therefore this will require more care.

In fact, this is actually a good thing to test! We should have tests where the callback:
* returns something that's an iterator but not an iterable
* returns something that's an iterable but not an iterator (so we have to get one through its Symbol.iterator method)
* returns something that's both an iterable and an iterator (generators fall into this category)
* returns something that's neither an iterator nor an iterable (should error, I'm guessing)

Reply:
@miorel I recently made a push addressing these changes, but i had an issue with the following test cases

* returns something that's an iterator but not an iterable
* returns something that's an iterable but not an iterator (so we have to get one through its Symbol.iterator method)

Reply:
Can you say more? Are you asking for what would be examples of those?

Reply:
> Can you say more? Are you asking for what would be examples of those?

Yeah I have two test for these cases, attempting for those.
```Typescript
  it("callback returns an iterator but not an iterable", () => {
    const iterator = [1, 2, 3].values();
    const callback = (x: unknown) =>
      ({
        next: () => ({ value: x, done: true }),
      }) as Iterator<unknown>;
    const flatMapResult = iterator.flatMap(callback);
    expect([...flatMapResult]).toStrictEqual([1, 2, 3]);
  });

  it("callback returns an iterable but not an iterator", () => {
    const iterator = [1, 2, 3].values();
    const callback = (x: unknown) =>
      ({
        [Symbol.iterator]: function* () {
          yield x;
        },
      }) as unknown as Iterator<unknown>;
    const flatMapResult = iterator.flatMap(callback);
    expect([...flatMapResult]).toStrictEqual([1, 2, 3]);
  });
  ```
  
  but I keep running into this error 
  
  
![image](https://github.com/user-attachments/assets/7b0e8d7f-3b5d-41de-ad1f-b6212aa23507)

also since this can return an Iterator or an Iterable or both, should I change the function signature to the snippet below. This also gives me a lot more errors too 😨 

```Typescript
callbackFn: (element: T, index: number) => Iterator<TOut> | Iterable<TOut>,
```

Reply:
As I'm reading `callbackFn: (element: T, index: number) => Iterator<TOut> | Iterable<TOut>` it's actually not immediately obvious what the `|` applies to. Is it `(() => A) | B` or `() => (A | B)`? So we may need to add parentheses around `(Iterator<TOut> | Iterable<TOut>)` in case that's an issue.

It might also be ok to use some type casts in the implementation if needed, maybe something like:

```ts
yield* (res[Symbol.iterator] ? res as Iterable<TOut> : (res as Iterator<TOut>).toIterable());
```

Basically we treat `res` as an iterable if it has a `Symbol.iterator` otherwise we assume it's an iterator, and convert it to an iterable. Give that a shot and let me know if it works?

Reply:
@miorel I'm getting `res.toIterable is not a function`

![image](https://github.com/user-attachments/assets/faaf34a3-131c-4c0f-b7ca-2479d5d21a62)

Reply:
Let's find sometime to pair on this live?

Reply:
Fixed Thank you! @miorel it was a lot of fun.